monitoring/tests/test_collectors.py

```diff
@@ -102,5 +106,144 @@ def test_stages_stats(self):
             self.assert_stage_stats_is_valid(build_stats.stages[1], self.default_stages_stats_response[1])
 
 
+class TestXmlReportCollector(unittest.TestCase):
+    def test_raises_error_when_json_does_not_exist(self):
+        with requests_mock.mock() as m:
+            m.get('http://1.2.3.4/build/1/test.json', status_code=404)
+
+            collector = XmlReportCollector(url='http://1.2.3.4/build/1/test.json')
+            with self.assertRaises(InvalidResponseCodeError):
+                collector.collect()
+
+    def test_raises_error_when_json_has_invalid_format(self):
+        with requests_mock.mock() as m:
+            response_text = 'Invalid JSON'
+            m.get('http://1.2.3.4/build/1/test.json', text=response_text)
+
+            collector = XmlReportCollector(url='http://1.2.3.4/build/1/test.json')
+            with self.assertRaises(JSONDecodeError):
+                collector.collect()
+
+    def test_raises_error_when_json_has_invalid_structure(self):
+        with requests_mock.mock() as m:
+            response_text = '{}'
+            m.get('http://1.2.3.4/build/1/test.json', text=response_text)
+
+            collector = XmlReportCollector(url='http://1.2.3.4/build/1/test.json')
+            with self.assertRaises(InvalidJsonFormatError):
+                collector.collect()
+
+    def test_raises_error_when_xml_reports_empty(self):
+        with requests_mock.mock() as m:
+            response_text = '{ "xml_reports": [], "html_report": "" }'
+            m.get('http://1.2.3.4/build/1/test.json', text=response_text)
+
+            collector = XmlReportCollector(url='http://1.2.3.4/build/1/test.json')
+            with self.assertRaises(EmptyXmlReportsListError):
+                collector.collect()
+
+    def test_raises_error_when_stats_do_not_exist(self):
+        with requests_mock.mock() as m:
+            response_text = '{ "xml_reports": ["test1.xml", "test2.xml"], "html_report": "report.html" }'
+            m.get('http://1.2.3.4/build/1/test.json', text=response_text)
+            m.get('http://1.2.3.4/build/1/test1.xml', text='')
+            m.get('http://1.2.3.4/build/1/test2.xml', status_code=404)
+
+            collector = XmlReportCollector(url='http://1.2.3.4/build/1/test.json')
+            with self.assertRaises(InvalidResponseCodeError):
+                collector.collect()
+
+    def test_raises_error_when_some_fields_do_not_exist(self):
+        with requests_mock.mock() as m:
+            response_text = '{ "xml_reports": ["report.xml"], "html_report": "report.html" }'
+            m.get('http://1.2.3.4/build/1/test.json', text=response_text)
```

Comment:
Multiple constant URLs (e.g. 'http://1.2.3.4/build/1/test.json') are repeated in the code.